aten/src/ATen/native/cuda/MaxUnpooling.cu

```diff
@@ -0,0 +1,590 @@
+    const int64_t outputWidth,
+    T* output) {
+  CUDA_KERNEL_LOOP(linearIndex, numInputElements) {
+    int c = (linearIndex / inputWidth / inputHeight) % numChannels;
+    int n = linearIndex / inputWidth / inputHeight / numChannels;
+    output += (n * numChannels + c) * outputHeight * outputWidth;
+    int maxind = indices[linearIndex];
+    output[maxind] = input[linearIndex];
+  }
+}
+
+template <typename T>
+__global__ void max_unpooling3d_forward_kernel(
+    PackedTensorAccessor<T, 4> input,
+    PackedTensorAccessor<int64_t, 4> indices,
+    T* output,
+    const int64_t oT,
+    const int64_t oH,
+    const int64_t oW,
+    const int64_t offsetZ) {
+  int64_t iColumn = blockIdx.x * blockDim.x + threadIdx.x;
+  int64_t iRow = blockIdx.y * blockDim.y + threadIdx.y;
+  int64_t iFrame = (blockIdx.z + offsetZ) % input.size(1); // input frame/time
+  int64_t slice = (blockIdx.z + offsetZ) / input.size(1); // input slice/feature
+  if (iRow < input.size(2) && iColumn < input.size(3)) {
+    T val = input[slice][iFrame][iRow][iColumn];
+    int64_t index = indices[slice][iFrame][iRow][iColumn];
+    output[slice * oT * oH * oW + index] = val;
+  }
+}
+
+template <typename T>
+__global__ void max_unpooling2d_backward_kernel(
+    const int64_t numInputElements,
+    const T* input,
+    const int64_t* indices,
+    const int64_t numChannels,
+    const int64_t inputHeight,
+    const int64_t inputWidth,
+    const int64_t outputHeight,
+    const int64_t outputWidth,
+    T* output) {
+  CUDA_KERNEL_LOOP(linearIndex, numInputElements) {
+    int c = (linearIndex / inputWidth / inputHeight) % numChannels;
+    int n = linearIndex / inputWidth / inputHeight / numChannels;
+    input += (n * numChannels + c) * outputHeight * outputWidth;
+    int maxind = indices[linearIndex];
+    output[linearIndex] = input[maxind];
+  }
+}
+
+template <typename T>
+__global__ void max_unpooling3d_backward_kernel(
+    T* gradOutputData,
+    int64_t oT,
+    int64_t oH,
+    int64_t oW,
+    PackedTensorAccessor<int64_t, 4> indices,
+    PackedTensorAccessor<T, 4> gradInput,
+    int offsetZ) {
+  int iColumn = blockIdx.x * blockDim.x + threadIdx.x;
+  int iRow = blockIdx.y * blockDim.y + threadIdx.y;
+  int iFrame = (blockIdx.z + offsetZ) % gradInput.size(1); // output frame/time
+  int slice =
+      (blockIdx.z + offsetZ) / gradInput.size(1); // output slice/feature
+
+  if (iRow < gradInput.size(2) && iColumn < gradInput.size(3)) {
+    int64_t index = indices[slice][iFrame][iRow][iColumn];
+    T grad_val = gradOutputData[slice * oT * oH * oW + index];
+    gradInput[slice][iFrame][iRow][iColumn] = grad_val;
+  }
+}
+
+Tensor& max_unpooling2d_forward_out_cuda(
+    Tensor& output,
+    const Tensor& self_,
+    const Tensor& indices_,
+    IntList output_size) {
+  AT_CHECK(
+      indices_.scalar_type() == at::ScalarType::Long,
+      "elements in indices should be type int64");
+  auto owidth = output_size[0];
+  auto oheight = output_size[1];
```

Comment:
I'm sorry but I gave you bad guidance here. `output_size[0]` is height, `output_size[1]` is width. The THNN ordering here is not the right thing to respect; instead we need to respect the documentation. I wrote some more guidance in the porting guide, which I've reproduced below:

Height-width argument order

Some kernels in THNN have very strange argument order; for example:

```
void THNN_(SpatialFractionalMaxPooling_updateOutput)(
           THCState *state,
           THCTensor *input,
           THCTensor *output,
           int outputW, int outputH,
           int poolSizeW, int poolSizeH,
           THCIndexTensor *indices,
           THCTensor *randomSamples)
```

outputW comes before outputH, even though conventional NCHW ordering says that height comes before width. *There is code generation logic specifically for THNN that reorders output_sizes to match the THNN kernel; you should manually reorder when you port.* So for example, the above arguments become:

```
void fractional_max_pool2d_out(
  Tensor& output,
  Tensor& indices,
  const Tensor& input,
  IntList pool_size,
  IntList output_size,
  const Tensor& randomSamples) {
  ...
  auto outputH = output[0]; // Notice they are reordered
  auto outputW = output[1];
  auto poolSizeH = pool_size[0];
  auto poolSizeW = pool_size[1];
}
```

The tests are supposed to catch if you've gotten it wrong, but we may have missed some cases, so be vigilant. In general, height comes before width. Here is a full list of dimension offsets (from aten/src/ATen/nn_parse.py), where negative numbers mean that you index from the end of the list:

```
DIMENSION_OFFSET = {
    'width': -1,
    'height': -2,
    'B': 0,
    'C': 1,
    'W': -1,
    'H': -2,
    'T': -3,
    'left': 0,
    'right': 1,
    'top': 2,
    'bottom': 3,
    'front': 4,
    'back': 5,
}
```

Reply:
Thanks for the heads up! I've pushed changed that assume that IntLists are either (T, H, W) or (H, W).